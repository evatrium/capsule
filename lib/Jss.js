(function (global, factory) {
    if (typeof define === "function" && define.amd) {
        define(['exports', 'react', 'react-jss', 'jss-preset-default', 'jss', 'react-jss/lib/JssProvider'], factory);
    } else if (typeof exports !== "undefined") {
        factory(exports, require('react'), require('react-jss'), require('jss-preset-default'), require('jss'), require('react-jss/lib/JssProvider'));
    } else {
        var mod = {
            exports: {}
        };
        factory(mod.exports, global.react, global.reactJss, global.jssPresetDefault, global.jss, global.JssProvider);
        global.Jss = mod.exports;
    }
})(this, function (exports, _react, _reactJss, _jssPresetDefault, _jss, _JssProvider) {
    'use strict';

    Object.defineProperty(exports, "__esModule", {
        value: true
    });
    exports.Jss = undefined;

    var _react2 = _interopRequireDefault(_react);

    var _reactJss2 = _interopRequireDefault(_reactJss);

    var _jssPresetDefault2 = _interopRequireDefault(_jssPresetDefault);

    var _JssProvider2 = _interopRequireDefault(_JssProvider);

    function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj : {
            default: obj
        };
    }

    var _jsxFileName = 'src/Jss.js';

    function _classCallCheck(instance, Constructor) {
        if (!(instance instanceof Constructor)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }

    var _createClass = function () {
        function defineProperties(target, props) {
            for (var i = 0; i < props.length; i++) {
                var descriptor = props[i];
                descriptor.enumerable = descriptor.enumerable || false;
                descriptor.configurable = true;
                if ("value" in descriptor) descriptor.writable = true;
                Object.defineProperty(target, descriptor.key, descriptor);
            }
        }

        return function (Constructor, protoProps, staticProps) {
            if (protoProps) defineProperties(Constructor.prototype, protoProps);
            if (staticProps) defineProperties(Constructor, staticProps);
            return Constructor;
        };
    }();

    function _possibleConstructorReturn(self, call) {
        if (!self) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }

        return call && (typeof call === "object" || typeof call === "function") ? call : self;
    }

    function _inherits(subClass, superClass) {
        if (typeof superClass !== "function" && superClass !== null) {
            throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
        }

        subClass.prototype = Object.create(superClass && superClass.prototype, {
            constructor: {
                value: subClass,
                enumerable: false,
                writable: true,
                configurable: true
            }
        });
        if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
    }

    var jss = (0, _jss.create)((0, _jssPresetDefault2.default)());

    var Jss = exports.Jss = function (_React$PureComponent) {
        _inherits(Jss, _React$PureComponent);

        function Jss() {
            _classCallCheck(this, Jss);

            return _possibleConstructorReturn(this, (Jss.__proto__ || Object.getPrototypeOf(Jss)).apply(this, arguments));
        }

        _createClass(Jss, [{
            key: 'render',
            value: function render() {
                var _props = this.props,
                    children = _props.children,
                    global_styles = _props.global_styles,
                    theme = _props.theme;

                var GlobalStylesWrapper = function GlobalStylesWrapper(props) {
                    return props.children;
                };
                var gs = global_styles ? global_styles : function () {
                    return {};
                };
                GlobalStylesWrapper = (0, _reactJss2.default)(global_styles)(GlobalStylesWrapper);
                return _react2.default.createElement(
                    _JssProvider2.default,
                    { jss: jss, __source: {
                            fileName: _jsxFileName,
                            lineNumber: 23
                        },
                        __self: this
                    },
                    _react2.default.createElement(
                        _reactJss.ThemeProvider,
                        { theme: theme ? theme : {}, __source: {
                                fileName: _jsxFileName,
                                lineNumber: 24
                            },
                            __self: this
                        },
                        _react2.default.createElement(
                            GlobalStylesWrapper,
                            {
                                __source: {
                                    fileName: _jsxFileName,
                                    lineNumber: 25
                                },
                                __self: this
                            },
                            children
                        )
                    )
                );
            }
        }]);

        return Jss;
    }(_react2.default.PureComponent);
});