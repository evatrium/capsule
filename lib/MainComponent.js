(function (global, factory) {
    if (typeof define === "function" && define.amd) {
        define(['exports', 'react', 'react-redux', 'react-jss', 'jss-preset-default', 'jss', 'react-jss/lib/JssProvider'], factory);
    } else if (typeof exports !== "undefined") {
        factory(exports, require('react'), require('react-redux'), require('react-jss'), require('jss-preset-default'), require('jss'), require('react-jss/lib/JssProvider'));
    } else {
        var mod = {
            exports: {}
        };
        factory(mod.exports, global.react, global.reactRedux, global.reactJss, global.jssPresetDefault, global.jss, global.JssProvider);
        global.MainComponent = mod.exports;
    }
})(this, function (exports, _react, _reactRedux, _reactJss, _jssPresetDefault, _jss, _JssProvider) {
    'use strict';

    Object.defineProperty(exports, "__esModule", {
        value: true
    });
    exports.MainComponent = undefined;

    var _react2 = _interopRequireDefault(_react);

    var _reactJss2 = _interopRequireDefault(_reactJss);

    var _jssPresetDefault2 = _interopRequireDefault(_jssPresetDefault);

    var _JssProvider2 = _interopRequireDefault(_JssProvider);

    function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj : {
            default: obj
        };
    }

    var _jsxFileName = 'src/MainComponent.js';

    function _classCallCheck(instance, Constructor) {
        if (!(instance instanceof Constructor)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }

    var _createClass = function () {
        function defineProperties(target, props) {
            for (var i = 0; i < props.length; i++) {
                var descriptor = props[i];
                descriptor.enumerable = descriptor.enumerable || false;
                descriptor.configurable = true;
                if ("value" in descriptor) descriptor.writable = true;
                Object.defineProperty(target, descriptor.key, descriptor);
            }
        }

        return function (Constructor, protoProps, staticProps) {
            if (protoProps) defineProperties(Constructor.prototype, protoProps);
            if (staticProps) defineProperties(Constructor, staticProps);
            return Constructor;
        };
    }();

    function _possibleConstructorReturn(self, call) {
        if (!self) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }

        return call && (typeof call === "object" || typeof call === "function") ? call : self;
    }

    function _inherits(subClass, superClass) {
        if (typeof superClass !== "function" && superClass !== null) {
            throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
        }

        subClass.prototype = Object.create(superClass && superClass.prototype, {
            constructor: {
                value: subClass,
                enumerable: false,
                writable: true,
                configurable: true
            }
        });
        if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
    }

    var jss = (0, _jss.create)((0, _jssPresetDefault2.default)());

    var MainComponent = exports.MainComponent = function (_React$PureComponent) {
        _inherits(MainComponent, _React$PureComponent);

        function MainComponent() {
            _classCallCheck(this, MainComponent);

            return _possibleConstructorReturn(this, (MainComponent.__proto__ || Object.getPrototypeOf(MainComponent)).apply(this, arguments));
        }

        _createClass(MainComponent, [{
            key: 'render',
            value: function render() {
                var _props = this.props,
                    children = _props.children,
                    store = _props.store,
                    global_styles = _props.global_styles,
                    theme = _props.theme;

                var GlobalStylesWrapper = function GlobalStylesWrapper(props) {
                    return props.children;
                };
                GlobalStylesWrapper = (0, _reactJss2.default)(global_styles)(GlobalStylesWrapper);
                return _react2.default.createElement(
                    _reactRedux.Provider,
                    { store: store, __source: {
                            fileName: _jsxFileName,
                            lineNumber: 21
                        },
                        __self: this
                    },
                    _react2.default.createElement(
                        _JssProvider2.default,
                        { jss: jss, __source: {
                                fileName: _jsxFileName,
                                lineNumber: 22
                            },
                            __self: this
                        },
                        _react2.default.createElement(
                            _reactJss.ThemeProvider,
                            { theme: theme, __source: {
                                    fileName: _jsxFileName,
                                    lineNumber: 23
                                },
                                __self: this
                            },
                            _react2.default.createElement(
                                GlobalStylesWrapper,
                                {
                                    __source: {
                                        fileName: _jsxFileName,
                                        lineNumber: 24
                                    },
                                    __self: this
                                },
                                children
                            )
                        )
                    )
                );
            }
        }]);

        return MainComponent;
    }(_react2.default.PureComponent);
});