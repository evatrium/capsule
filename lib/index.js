(function (global, factory) {
    if (typeof define === "function" && define.amd) {
        define(['exports', 'react', './redux-assist', '@iosio/components/lib/Jss'], factory);
    } else if (typeof exports !== "undefined") {
        factory(exports, require('react'), require('./redux-assist'), require('@iosio/components/lib/Jss'));
    } else {
        var mod = {
            exports: {}
        };
        factory(mod.exports, global.react, global.reduxAssist, global.Jss);
        global.index = mod.exports;
    }
})(this, function (exports, _react, _reduxAssist, _Jss) {
    'use strict';

    Object.defineProperty(exports, "__esModule", {
        value: true
    });
    exports.Capsule = exports.CapsuleProvider = exports.CapsuleModule = undefined;

    var _react2 = _interopRequireDefault(_react);

    function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj : {
            default: obj
        };
    }

    function _defineProperty(obj, key, value) {
        if (key in obj) {
            Object.defineProperty(obj, key, {
                value: value,
                enumerable: true,
                configurable: true,
                writable: true
            });
        } else {
            obj[key] = value;
        }

        return obj;
    }

    var _jsxFileName = 'src/index.js';

    var _createClass = function () {
        function defineProperties(target, props) {
            for (var i = 0; i < props.length; i++) {
                var descriptor = props[i];
                descriptor.enumerable = descriptor.enumerable || false;
                descriptor.configurable = true;
                if ("value" in descriptor) descriptor.writable = true;
                Object.defineProperty(target, descriptor.key, descriptor);
            }
        }

        return function (Constructor, protoProps, staticProps) {
            if (protoProps) defineProperties(Constructor.prototype, protoProps);
            if (staticProps) defineProperties(Constructor, staticProps);
            return Constructor;
        };
    }();

    function _possibleConstructorReturn(self, call) {
        if (!self) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }

        return call && (typeof call === "object" || typeof call === "function") ? call : self;
    }

    function _inherits(subClass, superClass) {
        if (typeof superClass !== "function" && superClass !== null) {
            throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
        }

        subClass.prototype = Object.create(superClass && superClass.prototype, {
            constructor: {
                value: subClass,
                enumerable: false,
                writable: true,
                configurable: true
            }
        });
        if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
    }

    function _classCallCheck(instance, Constructor) {
        if (!(instance instanceof Constructor)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }

    var CapsuleModule = exports.CapsuleModule = function CapsuleModule() {
        var _this2 = this;

        _classCallCheck(this, CapsuleModule);

        var _ref = new _reduxAssist.StoreModule(),
            reduxAssist = _ref.reduxAssist,
            store = _ref.store,
            connect = _ref.connect,
            Provider = _ref.Provider;

        var _provide_to_props = {};
        var collective_logic = {};
        var state = {};

        var CapsuleProvider = function CapsuleProvider(config) {
            //provide_to_props, provide_to_logic, theme, global_styles
            if (config && config.provide_to_props) {
                _provide_to_props = config.provide_to_props;
            }

            if (config && config.provide_to_logic) {
                collective_logic = Object.assign({}, collective_logic, config.provide_to_logic);
            }

            return function (App) {
                if (!App) {
                    return null;
                }
                return function (_React$Component) {
                    _inherits(CapsuleProvided, _React$Component);

                    function CapsuleProvided() {
                        _classCallCheck(this, CapsuleProvided);

                        return _possibleConstructorReturn(this, (CapsuleProvided.__proto__ || Object.getPrototypeOf(CapsuleProvided)).apply(this, arguments));
                    }

                    _createClass(CapsuleProvided, [{
                        key: 'render',
                        value: function render() {
                            return _react2.default.createElement(
                                Provider,
                                {
                                    __source: {
                                        fileName: _jsxFileName,
                                        lineNumber: 30
                                    },
                                    __self: this
                                },
                                _react2.default.createElement(
                                    _Jss.Jss,
                                    {
                                        theme: config && config.theme ? config.theme : {},
                                        global_styles: config && config.global_styles ? config.global_styles : function () {
                                            return {};
                                        }, __source: {
                                            fileName: _jsxFileName,
                                            lineNumber: 31
                                        },
                                        __self: this
                                    },
                                    _react2.default.createElement(App, Object.assign({}, this.props, _provide_to_props, {
                                        __source: {
                                            fileName: _jsxFileName,
                                            lineNumber: 38
                                        },
                                        __self: this
                                    }))
                                )
                            );
                        }
                    }]);

                    return CapsuleProvided;
                }(_react2.default.Component);
            };
        };

        var Capsule = function Capsule(_ref2) {
            var styles = _ref2.styles,
                reducer_name = _ref2.reducer_name,
                initial_state = _ref2.initial_state,
                mapStateToProps = _ref2.mapStateToProps,
                logic_name = _ref2.logic_name,
                logic = _ref2.logic,
                mapLogicToProps = _ref2.mapLogicToProps;


            if (reducer_name && initial_state) {
                state = reduxAssist(reducer_name, initial_state);
            }

            var return_logic = void 0;

            if (logic) {
                return_logic = logic({
                    state: state,
                    store: store,
                    collective: function collective() {
                        return Object.assign({}, collective_logic);
                    }
                });
            }

            if (logic_name && return_logic) {
                collective_logic = Object.assign({}, collective_logic, _defineProperty({}, logic_name, return_logic));
            }

            return function (Component) {

                var logic_props = {};
                if (mapLogicToProps) {
                    logic_props = mapLogicToProps(collective_logic);
                }

                if (Component && styles) {
                    Component = (0, _Jss.withStyles)(styles)(Component);
                }

                if (Component) {

                    var WithCapsule = function WithCapsule(props) {
                        return _react2.default.createElement(Component, Object.assign({}, props, logic_props, _provide_to_props, {
                            __source: {
                                fileName: _jsxFileName,
                                lineNumber: 91
                            },
                            __self: _this2
                        }));
                    };

                    if (mapStateToProps) {
                        return connect(mapStateToProps, null)(WithCapsule);
                    } else {
                        return WithCapsule;
                    }
                } else {
                    return return_logic;
                }
            };
        };

        return { CapsuleProvider: CapsuleProvider, Capsule: Capsule };
    };

    var _ref3 = new CapsuleModule();

    var CapsuleProvider = _ref3.CapsuleProvider,
        Capsule = _ref3.Capsule;
    exports.CapsuleProvider = CapsuleProvider;
    exports.Capsule = Capsule;
});