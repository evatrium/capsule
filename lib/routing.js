(function (global, factory) {
    if (typeof define === "function" && define.amd) {
        define(['exports', 'react', 'react-router-dom', 'history', 'query-string', '@iosio/utils/lib/empty_checks', '@iosio/utils/lib/misc', './index'], factory);
    } else if (typeof exports !== "undefined") {
        factory(exports, require('react'), require('react-router-dom'), require('history'), require('query-string'), require('@iosio/utils/lib/empty_checks'), require('@iosio/utils/lib/misc'), require('./index'));
    } else {
        var mod = {
            exports: {}
        };
        factory(mod.exports, global.react, global.reactRouterDom, global.history, global.queryString, global.empty_checks, global.misc, global.index);
        global.routing = mod.exports;
    }
})(this, function (exports, _react, _reactRouterDom, _history, _queryString, _empty_checks, _misc, _index) {
    'use strict';

    Object.defineProperty(exports, "__esModule", {
        value: true
    });
    exports.Redirect = exports.Switch = exports.Route = exports.routingLogic = exports.RouteFader = exports.Router = undefined;

    var _react2 = _interopRequireDefault(_react);

    var _queryString2 = _interopRequireDefault(_queryString);

    function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj : {
            default: obj
        };
    }

    var _jsxFileName = 'src/routing.js';

    function _classCallCheck(instance, Constructor) {
        if (!(instance instanceof Constructor)) {
            throw new TypeError("Cannot call a class as a function");
        }
    }

    var _createClass = function () {
        function defineProperties(target, props) {
            for (var i = 0; i < props.length; i++) {
                var descriptor = props[i];
                descriptor.enumerable = descriptor.enumerable || false;
                descriptor.configurable = true;
                if ("value" in descriptor) descriptor.writable = true;
                Object.defineProperty(target, descriptor.key, descriptor);
            }
        }

        return function (Constructor, protoProps, staticProps) {
            if (protoProps) defineProperties(Constructor.prototype, protoProps);
            if (staticProps) defineProperties(Constructor, staticProps);
            return Constructor;
        };
    }();

    function _possibleConstructorReturn(self, call) {
        if (!self) {
            throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
        }

        return call && (typeof call === "object" || typeof call === "function") ? call : self;
    }

    function _inherits(subClass, superClass) {
        if (typeof superClass !== "function" && superClass !== null) {
            throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
        }

        subClass.prototype = Object.create(superClass && superClass.prototype, {
            constructor: {
                value: subClass,
                enumerable: false,
                writable: true,
                configurable: true
            }
        });
        if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
    }

    /*
        can negate the need to use redux with routing
        if utilizing the same history object when pushing new paths or search params.
        so instead of using BrowserRouter, just use Router and provide the same history
     */
    var history = (0, _history.createBrowserHistory)();

    var routing = {
        reducer_name: 'routing',
        initial_state: {
            transitioning: false
        },
        logic_name: 'routing',
        logic: function logic(_ref) {
            var state = _ref.state,
                collective = _ref.collective;

            // console.log('collective logic in routing', Object.keys(collective()));
            var goTo = function goTo(path, search) {
                if (search && path) {
                    history.push({
                        search: _queryString2.default.stringify(search),
                        pathname: path

                    });
                } else if (path && !search) {
                    history.push({ pathname: path });
                } else if (search && !path) {
                    history.push({ search: _queryString2.default.stringify(search) });
                }
            };
            var getParams = function getParams(string) {
                if (!string) {
                    string = history.location.search;
                }
                var params = (0, _misc.deepClone)(_queryString2.default.parse(string)); //check and see if they fixed this
                return (0, _empty_checks.isEmpty)(params) ? false : params;
            };

            var transTo = function transTo(path, search) {
                var time = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : 300;

                state.routing.set.transitioning(true);

                setTimeout(function () {
                    goTo(path, search);
                    state.routing.set.transitioning(false);
                }, time);
            };

            return {
                goTo: goTo,
                transTo: transTo,
                getParams: getParams,
                history: history
            };
        }
    };

    var routingLogic = (0, _index.Capsule)(routing)();

    // console.log('&&&&&&&&&&&&&&&& routingLogic', routingLogic)

    var router = function (_React$PureComponent) {
        _inherits(router, _React$PureComponent);

        function router() {
            _classCallCheck(this, router);

            return _possibleConstructorReturn(this, (router.__proto__ || Object.getPrototypeOf(router)).apply(this, arguments));
        }

        _createClass(router, [{
            key: 'render',
            value: function render() {
                return _react2.default.createElement(_reactRouterDom.Router, { history: this.props.logic.history, children: this.props.children, __source: {
                        fileName: _jsxFileName,
                        lineNumber: 71
                    },
                    __self: this
                });
            }
        }]);

        return router;
    }(_react2.default.PureComponent);

    var Router = (0, _index.Capsule)({ logic: function logic() {
            return { history: routingLogic.history };
        } })(router);

    var routeFader = function (_React$PureComponent2) {
        _inherits(routeFader, _React$PureComponent2);

        function routeFader() {
            _classCallCheck(this, routeFader);

            return _possibleConstructorReturn(this, (routeFader.__proto__ || Object.getPrototypeOf(routeFader)).apply(this, arguments));
        }

        _createClass(routeFader, [{
            key: 'render',
            value: function render() {
                var _props = this.props,
                    classes = _props.classes,
                    className = _props.className,
                    no_transition = _props.no_transition,
                    style = _props.style,
                    ms_delay = _props.ms_delay,
                    transitioning = _props.transitioning,
                    children = _props.children,
                    cx = _props.cx;

                return _react2.default.createElement(
                    'div',
                    {
                        className: cx(classes.root, transitioning && !no_transition ? null : classes.show, className),
                        style: Object.assign({}, style, {
                            transition: 'opacity ' + (ms_delay ? no_transition ? 0 : ms_delay : no_transition ? 0 : 300) + 'ms ease-in-out'
                        }), __source: {
                            fileName: _jsxFileName,
                            lineNumber: 82
                        },
                        __self: this
                    },
                    children
                );
            }
        }]);

        return routeFader;
    }(_react2.default.PureComponent);

    var RouteFader = (0, _index.Capsule)({
        styles: {
            root: {
                height: '100%',
                width: '100%',
                opacity: 0
            },
            show: {
                opacity: 1
            }
        },
        mapStateToProps: function mapStateToProps(state) {
            return {
                transitioning: state.routing.transitioning
            };
        }
    })(routeFader);

    exports.Router = Router;
    exports.RouteFader = RouteFader;
    exports.routingLogic = routingLogic;
    exports.Route = _reactRouterDom.Route;
    exports.Switch = _reactRouterDom.Switch;
    exports.Redirect = _reactRouterDom.Redirect;
});