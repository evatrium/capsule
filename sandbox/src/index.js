let log_this = '';
const log = (it) => {
    let val;
    if (typeof it !== 'string') {
        val = JSON.stringify(it, null, 4);
    } else {
        val = it;
    }
    log_this = '\n' + log_this + '\n' + val + '\n';
};


export const tester = () => {
    return new Promise((resolve, reject) => {
        console.log(log_this);

        console.log('working');


        resolve()
    })
};