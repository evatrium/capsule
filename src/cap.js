import {Capsule} from "./index";


const user = Capsule('user', {
    active: false,
    first_name: '',
    last_name: '',
    src: ''
});

const onUserStateChange = ({active}) => {
    user.merge({active});
};

const Avatar = user(({active, first_name, last_name, src}) => {

    return (
        <UserBox>
            <Img src={src}/>
            <Typo text={first_name + ' ' + last_name}/>
        </UserBox>
    );
});

