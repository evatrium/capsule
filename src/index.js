export {
    CapsuleProvider,
    Capsule,
    events,
    lazyLoader,
    routing,
    Router,
    Linkage,
    jss_nano,
    cn
} from './created';