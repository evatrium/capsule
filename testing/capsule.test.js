import React from 'react';
import renderer from 'react-test-renderer';
import {isShape} from '@iosio/utils/lib/isShape';


import {CapsuleModule} from "../src";

import {TestMain, BaseComponent} from "./capsule_test_components";


const theme = {
    color: {
        light: '#93ccdc',
        dark: '#252336'
    }
};

const global_styles = () => ({
    '@global': {
        '*': {
            boxSizing: 'border-box',
        },
    },
});


describe('CapsuleModule', () => {


    describe('Capsule', () => {


        describe('logic', () => {


            it('logic should return logic_data from called capsule if no component is passed', () => {

                const {CapsuleProvider, Capsule} = new CapsuleModule();

                CapsuleProvider()(TestMain);

                const capsule_logic = Capsule({
                    reducer_name: 'user',
                    initial_state: {
                        logged_in: false,
                        name: 'asdf'
                    },
                    logic: () => ({
                        test: () => 'value'
                    })

                })();//called with no component

                expect(capsule_logic.test()).toEqual('value');

            });


            it(`
            when assigned logic_name (namespace),
            logic should be accessible via collective in other capsules`, () => {

                const {CapsuleProvider, Capsule} = new CapsuleModule();

                CapsuleProvider()(TestMain);

                const capsule_logic1 = Capsule({
                    logic_name: 'tester1_namespace',
                    logic: ({collective}) => ({
                        test1_collective: () => {
                            const results = collective().tester2_namespace.test2();
                            // console.log(results)
                            expect(results).toBe('test2-value')
                        },
                        test1_func: () => {
                            return 'test1-value'
                        }
                    })

                })();

                const capsule_logic2 = Capsule({
                    logic_name: 'tester2_namespace',
                    logic: ({collective}) => ({
                        test2_collective: () => {
                            const results = collective().tester1_namespace.test1_func();
                            // console.log(results)
                            expect(results).toBe('test1-value')
                        },
                        test2: () => {
                            return 'test2-value'
                        }
                    })
                })();

                capsule_logic1.test1_collective();
                capsule_logic2.test2_collective();

            });


            it('verify that logic can be mapped to props. (must provide logic_name namespace)', () => {

                const {CapsuleProvider, Capsule} = new CapsuleModule();


                let Component = (props) => (
                    <div>
                        {Object.keys(props).map((prop, i) => (
                            <p key={i}> prop:{prop}</p>
                        ))}
                    </div>
                );

                let CapsuleWrappedComponent = Capsule({
                    reducer_name: 'user',
                    initial_state: {
                        logged_in: false,
                        name: null
                    },
                    logic_name: 'user',
                    logic: ({state}) => ({
                        login: () => {
                            //setting individually just for example
                            state.user.set.logged_in(true);
                            state.user.set.name('Joe Dirt')
                        },
                        logout: () => {
                            state.user.update((state) => ({
                                ...state,
                                logged_in: false,
                                name: null
                            }))
                        }
                    }),
                    mapLogicToProps: (logic) => ({
                        login: logic.user.login,
                        logout: logic.user.logout
                    })
                })(Component);


                const Main = (props) => (
                    <BaseComponent {...props}>
                        <CapsuleWrappedComponent/>
                    </BaseComponent>
                );

                const CapsuleProvidedApp = CapsuleProvider()(Main);

                const tree = renderer.create(
                    <CapsuleProvidedApp/>
                );

                expect(tree).toMatchSnapshot();


                //end Capsule
            });


        });//end describe logic


        describe(`
        redux-assist itself has already been tested.
        Validate that it works in this context.
        `, () => {


            it('should return state controller object of this shape inside of logic', () => {

                const {CapsuleProvider, Capsule} = new CapsuleModule();

                CapsuleProvider()(TestMain);

                Capsule({
                    reducer_name: 'user',
                    initial_state: {
                        logged_in: false,
                        name: 'asdf'
                    },
                    logic: ({state}) => {

                        expect(isShape(state, {
                            user: {
                                type: 'object',
                                required: true,
                                shape: {
                                    set: {
                                        required: true,
                                        type: 'object',
                                        shape: {
                                            logged_in: {
                                                required: true,
                                                type: 'function'
                                            },
                                            name: {
                                                required: true,
                                                type: 'function'
                                            }
                                        }
                                    },
                                    get: {
                                        required: true,
                                        type: 'object',
                                        shape: {
                                            logged_in: {
                                                required: true,
                                                type: 'function'
                                            },
                                            name: {
                                                required: true,
                                                type: 'function'
                                            }
                                        }
                                    },
                                    update: {
                                        required: true,
                                        type: 'function'
                                    }
                                }
                            }
                        }).ok).toBe(true);
                    }

                });
                //end Capsule
            });


            it('verify that state updates on the component and that state is mapped to props', () => {

                const {CapsuleProvider, Capsule} = new CapsuleModule();

                CapsuleProvider()(TestMain);

                let Component = (props) => (
                    <div>
                        {JSON.stringify(props, null, 4)}
                    </div>
                );

                let user_logic = Capsule({
                    reducer_name: 'user',
                    initial_state: {
                        logged_in: false,
                        name: null
                    },
                    logic: ({state}) => ({
                        login: () => {
                            //setting individually just for example
                            state.user.set.logged_in(true);
                            state.user.set.name('Joe Dirt')
                        },
                        logout: () => {
                            state.user.update((state) => ({
                                ...state,
                                logged_in: false,
                                name: null
                            }))
                        }
                    })
                })();

                let ConnectedComponent = Capsule({

                    mapStateToProps: (state) => ({
                        name: state.user.name,
                        logged_in: state.user.logged_in
                    })
                })(Component);

                const Main = (props) => (
                    <BaseComponent {...props}>
                        <ConnectedComponent/>
                    </BaseComponent>
                );

                const CapsuleProvidedApp = CapsuleProvider()(Main);

                const tree = renderer.create(
                    <CapsuleProvidedApp/>
                );

                expect(tree).toMatchSnapshot();

                user_logic.login();

                expect(tree).toMatchSnapshot();

                user_logic.logout();

                expect(tree).toMatchSnapshot();

                //end Capsule
            });


        });//end describe redux-assist


    });//end describe Capsule


    describe('CapsuleProvider', () => {

        it('should should render without crashing. no config, should still pass props', () => {

            const {CapsuleProvider, Capsule} = new CapsuleModule();

            const CapsuleProvidedApp = CapsuleProvider()(TestMain);

            const tree = renderer.create(
                <CapsuleProvidedApp test={"no config"}/>
            );

            expect(tree).toMatchSnapshot();


        });


        it('should provide to props.', () => {

            const {CapsuleProvider, Capsule} = new CapsuleModule();

            const CapsuleProvidedApp = CapsuleProvider({
                provide_to_props: {provide_to_props: 'asdf'}
            })(TestMain);

            const tree = renderer.create(
                <CapsuleProvidedApp/>
            );

            expect(tree).toMatchSnapshot();

        });


        it('should provide theme', () => {

            const {CapsuleProvider, Capsule} = new CapsuleModule();


            let ThemedComponent = (props) => (
                <div className={props}>
                    hello
                </div>
            );

            ThemedComponent = Capsule({
                styles: (theme) => ({
                    title: {
                        color: theme.color.light
                    }
                })
            })(ThemedComponent);

            const Main = (props) => (
                <BaseComponent {...props}>
                    <ThemedComponent/>
                </BaseComponent>
            );

            const CapsuleProvidedApp = CapsuleProvider({
                theme,
            })(Main);

            const tree = renderer.create(
                <CapsuleProvidedApp/>
            );


            expect(tree).toMatchSnapshot();

        });


        it('should provide test_logic to logic via collective ', () => {

            const {CapsuleProvider, Capsule} = new CapsuleModule();


            const CapsuleProvidedApp = CapsuleProvider({
                provide_to_logic: {
                    test_logic: 'test_value'
                }
            })(TestMain);


            let logic_provided_capsule = Capsule({
                logic: ({collective}) => ({
                    get_test_logic: () => {
                        return collective().test_logic
                    }
                })
            })();

            const value = logic_provided_capsule.get_test_logic();
            // console.log('value: ', value)

            expect(value).toEqual('test_value')

        });

    });


});
