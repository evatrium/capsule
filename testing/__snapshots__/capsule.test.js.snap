// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CapsuleModule Capsule 
        redux-assist itself has already been tested.
        Validate that it works in this context.
         verify that state updates on the component and that state is mapped to props 1`] = `
<div>
  {
    "name": null,
    "logged_in": false
}
</div>
`;

exports[`CapsuleModule Capsule 
        redux-assist itself has already been tested.
        Validate that it works in this context.
         verify that state updates on the component and that state is mapped to props 2`] = `
<div>
  {
    "name": "Joe Dirt",
    "logged_in": true
}
</div>
`;

exports[`CapsuleModule Capsule 
        redux-assist itself has already been tested.
        Validate that it works in this context.
         verify that state updates on the component and that state is mapped to props 3`] = `
<div>
  {
    "name": null,
    "logged_in": false
}
</div>
`;

exports[`CapsuleModule Capsule logic verify that logic can be mapped to props. (must provide logic_name namespace) 1`] = `
<div>
  <p>
     prop:
    login
  </p>
  <p>
     prop:
    logout
  </p>
</div>
`;

exports[`CapsuleModule CapsuleProvider should provide theme 1`] = `
<div
  className={
    Object {
      "classes": Object {
        "title": "ThemedComponent-title-0-2-1",
      },
      "theme": Object {
        "color": Object {
          "dark": "#252336",
          "light": "#93ccdc",
        },
      },
    }
  }
>
  hello
</div>
`;

exports[`CapsuleModule CapsuleProvider should provide to props. 1`] = `
<div>
  {"provide_to_props":"asdf"}
</div>
`;

exports[`CapsuleModule CapsuleProvider should should render without crashing. no config, should still pass props 1`] = `
<div>
  {"test":"no config"}
</div>
`;
